### ISSUE TITLE
Gateway timeout on 200k images

### ISSUE THREAD

**Issue by nagem07** [exadel-inc/CompreFace#776]
**Describe the bug**
After adding 200,000 images through Python SDK recognition service does not work anymore and returns 504 gateway timeout error.

**To Reproduce**
1. Add 200k images
2. Go to UI and try recognition service

**Expected behavior**
Person is recognised

**Screenshots**

**Desktop (please complete the following information):**
 - Ubuntu 20.04
 - Chrome
 - CompreFace 1.0.0
 - RTX2060
 - 64GB RAM

**Additional context**
Logs (Using external host as files are too big)
[https://www.mediafire.com/file/h2gap9jn3cja01i/DB.log/file](DB Log)
[https://www.mediafire.com/file/sl76j63by5f0xcl/Admin.log/file](Admin Log)
[https://www.mediafire.com/file/d4jumgd0b609ssl/FE.log/file](FE Log)
[https://www.mediafire.com/file/ibvtf15kdvaq10l/API.log/file](API Log)
[https://www.mediafire.com/file/41z5mttlld1us0n/Core.log/file](Core Log)

**Comment by pospielov** [comment]
Hi, sorry for so long response, I was on a long vacation.
According to your API.log file, java fails because it lacks memory. Did you update the .env file? There is such an option in the default configuration:
`compreface_api_java_options=-Xmx4g`
You should update it, e.g.:
`compreface_api_java_options=-Xmx16g`

**Comment by nagem07** [comment]
Hey,

Yeah I did update the .env file and it fixed the issue, however even if I set it to 16gb on 180k images it shoots up to 28GB, correct me if I am wrong but if I was to use a million images we would be looking at 256 - 384GB of RAM?

**Comment by pospielov** [comment]
yes, unfortunately, CompreFace is not optimized for such huge face collections. 
This will require different approaches which could be too heavy for small collections.
So we had to choose which collection size to support and for now we chose small collections. 
I believe 100k is the max comfortable size for CompreFace, of course, it will work with bigger collections, but as you mentioned it would cost too many resources.

**Comment by nagem07** [comment]
I see you prioritized smaller collections. From your knowledge, what would be the steps to optimize CompreFace for bigger collections?
I have also noticed very slow service initialization, on a 180k images. Meaning that after a reboot, system takes at least 30 minutes of time to load images back into the RAM. Could that be addressed as well?

**Comment by nagem07** [comment]
Thank you for your answer and sorry for the delay in getting back to you. So having had a look at Milvus, I understand it will be situated before Postgres in the architecture, so embeddings will be saved in Milvus, while subject information would be on Postgres. Am I correct?

Having said that could you advise on which scripts will require changes in order to integrate Milvus? I am looking at above a million images, and using the current architecture has a high cost in terms of resources.

**Comment by pospielov** [comment]
I didn't research it deeply. I think yes, it should be similar to your description.
You need to replace this class, and probably lots of logic related to it with Milvus logic:
https://github.com/exadel-inc/CompreFace/blob/master/java/api/src/main/java/com/exadel/frs/core/trainservice/component/classifiers/EuclideanDistanceClassifier.java

**Comment by martinenkoEduard** [comment]
What is the maxium practical amount of faces?

**Comment by nagem07** [comment]
> I didn't research it deeply. I think yes, it should be similar to your description. You need to replace this class, and probably lots of logic related to it with Milvus logic: https://github.com/exadel-inc/CompreFace/blob/master/java/api/src/main/java/com/exadel/frs/core/trainservice/component/classifiers/EuclideanDistanceClassifier.java

Would that be the sole class that requires modification or are there others as well?

**Comment by pospielov** [comment]
Others as well, this is just where you can start discovering.
